### PR TITLE
Read Mavlink from CSV file. Modules completed with tests

### PR DESCRIPTION
**Pull request by janez45**

### CODE REVIEW

**Comment by janez45** [comment]
Completed my task with tests written. My main concern is if the pilots make a mistake with the params. Like if a value that should be zero is not set to zero (e.g. param2 for waypoint). I wonder if there should be more safeguarding in that sense?

**Comment by janez45** [comment]
Another thing I'm concerned about is whether the code should crash if they forget a command parameter. Say, a 0 is missed.  I don't think they'd want a drone to accidentally skip a command without warning. Gonna actually make that a crash

File: modules/advanced_csv_to_commands.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    "loiter_timed": dronekit.mavutil.mavlink.MAV_CMD_NAV_LOITER_TIME,
+    "loiter_unlimited": dronekit.mavutil.mavlink.MAV_CMD_NAV_LOITER_UNLIM,
+    "do_jump": dronekit.mavutil.mavlink.MAV_CMD_DO_JUMP,
+}
+
+
+def generate_command_advanced(
+    frame: str,
+    command_type: str,
+    param1: float,  # I'm anxious about this because of do_jump which requires ints
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:30]: This function can dispatch to more specialized functions to construct the various commands.

File: modules/advanced_csv_to_commands.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    param2: float,
+    param3: float,
+    param4: float,
+    param5: float,
+    param6: float,
+    param7: float,
+) -> "tuple[bool,dronekit.Command]":
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:37]: `"tuple[bool, dronekit.Command | None]"`

File: modules/advanced_csv_to_commands.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    command_type: The command type. Depending on this value, params 1 through 7 will take on different meanings. This should be one of the following enums:
+        - land:
+        - return_to_launch: Return to the home location where the vehicle was last armed.
+        - takeoff
+        - waypoint: Navigate to the specified waypoint
+        - waypoint_spline
+        - loiter_timed: Loiter at the specified location for an specified amount of time
+        - loiter_unlimited: Loiter at the specified location for an unlimited amount of time
+        - do_jump: Jump to the specified command in the mission list.
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:53]: I think linking the Confluence document is sufficient, no need to repeat the information from the dictionary above or the information in the document.

File: modules/advanced_csv_to_commands.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    A method that reads a list of advanced commands from a csv file and generates a mission.
+    Parameters:
+        - mission_file_path: The advanced csv file containing the different mission commands
+
+    Returns:
+        - Whether or not the mission was successfully created
+        - A list of dronekit commands that represent the mission
+
+    """
+    # Does the file path exist?
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:102]: Remove.

File: modules/advanced_csv_to_commands.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    if not mission_file_path.exists():
+        return False, None
+
+    mission = []
+    with open(mission_file_path, encoding="utf-8") as file:
+        for line in file:
+            # Skip header and empty lines
+            parts = line.split(",")
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:110]: Move below the conditional below.

File: modules/advanced_csv_to_commands.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            if line in "frame,command_type,param1,param2,param3,param4,param5,param6,param7\n":
+                continue
+
+            # If the parameters are messed up
+            if len(parts) != 9:
+                return False, None
+
+            frame, command_type, param1, param2, param3, param4, param5, param6, param7 = parts
+            success, command = generate_command_advanced(
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:119]: `result, command =`

File: modules/advanced_csv_to_commands.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                command_type,
+                float(param1),
+                float(param2),
+                float(param3),
+                float(param4),
+                float(param5),
+                float(param6),
+                float(param7),
+            )
+            # Was the command successfully generated?
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:130]: Remove.

File: modules/advanced_csv_to_commands.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            if not success:
+                return False, None
+
+            mission.append(command)
+
+    if len(mission) > 0:
+        return True, mission
+
+    return False, None
```
Comment (Xierumeng) [inline modules/advanced_csv_to_commands.py:139]: Switch:
```py
if len(mission) == 0:
    return False, None

return True, mission
```

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: modules/advanced_csv_to_commands.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+"""
+The input matrix maps the command types to the valid parameters.
+See documentation for more information: https://uwarg-docs.atlassian.net/wiki/spaces/CV/pages/2567274629/Encoded+MAVLink+commands 
+0: Parameter should be set to zero
+1: Parameter should be a float
+2: Parameter should be an integer
+"""
+COMMAND_TO_PARAMETER_MATRIX = {
```
Comment (AaronWang04) [inline modules/advanced_csv_to_commands.py:33]: this could also be ENUMs

**Review by AaronWang04 — APPROVED** [review]
lgtm, nice job!

**Pull request merged**